### PR TITLE
DPP 0.23 Upgrade

### PR DESCRIPTION
**Pull request by HashEngineering**

### CODE REVIEW

File: src/main/kotlin/org/dashj/platform/dpp/statetransition/StateTransition.kt [hunk 1]
```diff
@@ -101,8 +111,26 @@ abstract class StateTransition(
...
+            IdentityPublicKey.Type.ECDSA_HASH160 -> {
+                val privateKeyModel = ECKey.fromPrivate(privateKey)
+                signature = privateKeyModel.signHash(Sha256Hash.wrap(data))
+            }
+            IdentityPublicKey.Type.BLS12_381 -> {
+                val privateKeyModel = BLSSecretKey(privateKey)
+                val blsSignature = privateKeyModel.Sign(Sha256Hash.wrap(data))
+
+                signature = blsSignature.bitcoinSerialize()
+            }
```
Comment (HashEngineering) [inline src/main/kotlin/org/dashj/platform/dpp/statetransition/StateTransition.kt:131]: One big change is adding support for BLS public keys in the identity for digital signatures (sign / verify)

Comment (HashEngineering) [reply src/main/kotlin/org/dashj/platform/dpp/statetransition/StateTransition.kt:131]: DashPay won't be supporting BLS in identities for the time being.

File: src/main/kotlin/org/dashj/platform/dpp/statetransition/StateTransitionIdentitySigned.kt [hunk 2]
```diff
@@ -137,28 +152,43 @@ abstract class StateTransitionIdentitySigned(
...
+
     /**
      * Returns minimal key security level that can be used to sign this ST.
      * Override this method if the ST requires a different security level.
      *
      * @return {number}
      */
     open fun getKeySecurityLevelRequirement(): IdentityPublicKey.SecurityLevel {
-        return IdentityPublicKey.SecurityLevel.MASTER
+        return IdentityPublicKey.SecurityLevel.HIGH
```
Comment (HashEngineering) [inline src/main/kotlin/org/dashj/platform/dpp/statetransition/StateTransitionIdentitySigned.kt:192]: This was a big change in 0.23

Comment (HashEngineering) [reply src/main/kotlin/org/dashj/platform/dpp/statetransition/StateTransitionIdentitySigned.kt:192]: Identities will need a key with `HIGH` to sign state transitions (to create usernames, profiles, etc)

File: src/main/kotlin/org/dashj/platform/dpp/util/Entropy.kt [hunk 3]
```diff
@@ -31,12 +31,17 @@ object Entropy {
 
     @JvmStatic
     fun generate(): ByteArray {
-        return mockEntroy ?: secureRandom.generateSeed(SEED_SIZE)
+        if (mockEntropy == null) {
+            val bytes32 = ByteArray(32)
+            secureRandom.nextBytes(bytes32)
+            return bytes32
+        }
+        return mockEntropy!!
```
Comment (HashEngineering) [inline src/main/kotlin/org/dashj/platform/dpp/util/Entropy.kt:39]: Previously we were using `generateSeed` instead of `nextBytes` to get a random value.

**Comment by HashEngineering** [comment]
CI is failing because `dashj-core` was updated, but those changes haven't been published.

File: src/test/kotlin/org/dashj/platform/dpp/Fixtures.kt [hunk 4]
```diff
@@ -375,7 +365,7 @@ object Fixtures {
         return StateTransitionFactory(dpp, stateRepository).createStateTransition(rawContractST) as DataContractCreateTransition
     }
 
-    fun getDocumentsSTSignedFixture(): DocumentsBatchTransition {
+    /*fun getDocumentsSTSignedFixture(): DocumentsBatchTransition {
```
Comment (HashEngineering) [inline src/test/kotlin/org/dashj/platform/dpp/Fixtures.kt:368]: this is obsolete, remove it

File: src/main/kotlin/org/dashj/platform/dpp/identity/IdentityPublicKey.kt [hunk 5]
```diff
@@ -9,27 +9,32 @@ package org.dashj.platform.dpp.identity
...
 ) : BaseObject() {
 
-    enum class TYPES(val value: Int) {
+    enum class Type(val value: Int) {
         ECDSA_SECP256K1(0),
         BLS12_381(1),
-        ECDSA_HASH160(2);
+        ECDSA_HASH160(2),
+        BIP13_SCRIPT_HASH(3),
+        INVALID(30000); // for tests
```
Comment (HashEngineering) [inline src/main/kotlin/org/dashj/platform/dpp/identity/IdentityPublicKey.kt:33]: `BIP13_SCRIPT_HASH`

There aren't any sign or verify functions that support this type.

**Review by Syn-McJ — APPROVED** [review]
Looks great

**Pull request merged**